tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb

```diff
@@ -327,170 +408,200 @@
       ]
     },
     {
-      "cell_type": "code",
-      "execution_count": 0,
       "metadata": {
-        "colab": {},
+        "id": "3t5ibNo05jCB",
         "colab_type": "code",
-        "id": "2M7LmLtGEMQJ"
+        "colab": {}
       },
-      "outputs": [],
+      "cell_type": "code",
       "source": [
-        "def train(dataset, epochs, noise_dim):  \n",
-        "  for epoch in range(epochs):\n",
-        "    start = time.time()\n",
-        "    \n",
-        "    for images in dataset:\n",
-        "      # generating noise from a uniform distribution\n",
+        "def train_step(images):\n",
+        "   # generating noise from a uniform distribution\n",
         "      noise = tf.random_normal([BATCH_SIZE, noise_dim])\n",
         "      \n",
         "      with tf.GradientTape() as gen_tape, tf.GradientTape() as disc_tape:\n",
         "        generated_images = generator(noise, training=True)\n",
         "      \n",
         "        real_output = discriminator(images, training=True)\n",
         "        generated_output = discriminator(generated_images, training=True)\n",
-        "        \n",
+        "         \n",
         "        gen_loss = generator_loss(generated_output)\n",
         "        disc_loss = discriminator_loss(real_output, generated_output)\n",
         "        \n",
         "      gradients_of_generator = gen_tape.gradient(gen_loss, generator.variables)\n",
         "      gradients_of_discriminator = disc_tape.gradient(disc_loss, discriminator.variables)\n",
         "      \n",
         "      generator_optimizer.apply_gradients(zip(gradients_of_generator, generator.variables))\n",
-        "      discriminator_optimizer.apply_gradients(zip(gradients_of_discriminator, discriminator.variables))\n",
+        "      discriminator_optimizer.apply_gradients(zip(gradients_of_discriminator, discriminator.variables))"
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "metadata": {
+        "colab_type": "text",
+        "id": "6TSZgwc2BUQ-"
+      },
+      "cell_type": "markdown",
+      "source": [
         "\n",
-        "      \n",
-        "    if epoch % 1 == 0:\n",
-        "      display.clear_output(wait=True)\n",
-        "      generate_and_save_images(generator,\n",
+        "This model takes about ~30 seconds per epoch to train on a single Tesla K80 on Colab, as of October 2018. \n",
+        "\n",
+        "Eager execution can sometimes be slower than executing the equivalent graph due to overheads of interpreting Python code. By using [tf.contrib.eager.defun](https://www.tensorflow.org/api_docs/python/tf/contrib/eager/defun) to create graph functions, we get a ~20 secs/epoch performance boost (from ~50 secs/epoch down to ~30 secs/epoch). This way we get the best of both eager execution (easier for debugging) and graph mode (better performance)."
```

Comment:
I'd rephrase this a bit (so as to not blame just the Python interpreter). There may be other graph optimizations we're benefiting from here as well (constant folding, common subexpression elimination, layout changes etc.). Though, in this particular case it is probably the Python overhead (since we haven't enabled grappler passes on function yet). How about:

"Eager execution can be slower than executing the equivalent graph as it can't benefit from whole-program optimizations on the graph, and also incurs overheads of interpreting Python code. By using ..."

Reply:
Updated as suggested.